### PR TITLE
Use subflow per request

### PR DESCRIPTION
**Pull request by hseeberger**
This PR introduces a new approach intended to simplify using Streamee.

The core design idea is to run per-request substreams in the processors instead of embeding the process flow directly. This makes correlating requests with promised responses – either via sequence (fragile) or correlation IDs (cumbersome, boilerplate) – obsolete, because there exists just one request and promised response pair per substream.

Here's the relevant code from `Processor`:

``` scala
Source
  .queue[(A, Promise[B])](1, OverflowStrategy.dropNew) // No need to use a large buffer, beause a substream is run for each request!
  .toMat(Sink.foreach {
    case (request, response) =>
      response.completeWith(Source.single(request).via(process).runWith(Sink.head))
  })(Keep.both)
  .run()
```

Hence users simply can express processes in terms of domain logic and data and never have to worry about correlation or sequence. In particular it is possible to use flow operators like `mapAsyncUnordered` for fast parallelism or graphs which use fan-in and fan-out shapes.

Here's a simple example from `DemoProcessor`:

``` scala
final case class ValidateQuestionIn(question: String)
final case class ValidateQuestionOut(question: String)

def validateQuestion: Stage[ValidateQuestionIn, ValidateQuestionOut] =
  Flow[ErrorOr[ValidateQuestionIn]]
    .map(_.flatMap {
      case ValidateQuestionIn(question) =>
        if (question.isEmpty)
          Left(Error.EmptyQuestion)
        else
          Right(ValidateQuestionOut(question))
    })
```

Another advantage of this new design is that failure of the substream leads to failing the request fast: the client needs not to wait for the promised response to expire and the actual exception is shown in the log.

As far as I can tell, all so far features are still supported – including registering with Akka Coordinated Shutdown to ensure to not loose in-flight requests – except for one: there is no more per-endpoint backpressure like before, i.e. the application will accept as many requests as allowed by the Akka HTTP configuration (the default is 1024 connections).

The demo application – see `DemoProcessor` for details – has been updated with a more complex process that shows
- using `Either` to signal possible errors of process stages (no `split` and `combine`, but these would work perfectly)
- using `mapAsyncUnordered` for parallelism
- using non-trivial graphs including `Unzip` and `Merge` shapes
- random exceptions in stages

### CODE REVIEW

**Comment by hseeberger** [comment]
@aquamatthias @markusjura please see this PR (I actually wrote some text in the leading comment) with a new and simplified design approach for Streamee.

File: streamee/src/main/scala/io/moia/streamee/Processor.scala [hunk 1]
```diff
@@ -76,123 +47,73 @@ object Processor extends Logging {
...
+      process: Flow[A, B, Any],
+      timeout: FiniteDuration,
+      name: String
+  )(implicit ec: ExecutionContext, mat: Materializer, scheduler: Scheduler): Processor[A, B] =
+    new ProcessorImpl(process, timeout, name)
 }
 
 /**
   * Process requests with a running process or shut it down.
+  *     * responses.
```
Comment (aquamatthias) [inline streamee/src/main/scala/io/moia/streamee/Processor.scala]: Comment looks strange. intended?

File: streamee/src/test/scala/io/moia/streamee/ProcessorTests.scala [hunk 2]
```diff
@@ -28,135 +28,89 @@ object ProcessorTests extends ActorTestSuite {
...
-          .map { responses =>
-            assert(
-              responses == List(43,
-                                PromiseExpired(timeout, "43"),
-                                45,
-                                PromiseExpired(timeout, "45"))
-            )
-          }
-      }
+//      'drop - {
```
Comment (aquamatthias) [inline streamee/src/test/scala/io/moia/streamee/ProcessorTests.scala]: either remove or fix.

File: streamee/src/main/scala/io/moia/streamee/Processor.scala (hunk 1 above)
Comment (hseeberger) [reply streamee/src/main/scala/io/moia/streamee/Processor.scala]: No. Thanks.

File: streamee/src/test/scala/io/moia/streamee/ProcessorTests.scala (hunk 2 above)
Comment (hseeberger) [reply streamee/src/test/scala/io/moia/streamee/ProcessorTests.scala]: A leftover. Thanks. Removed.

**Comment by hseeberger** [comment]
@aquamatthias thanks for the review. I agree that it would be nice to offer both options. Let me see if I can unify the `Processor` API.

**Comment by hseeberger** [comment]
Updated the demo and tests. `permanent` and `perRequest` are well tested and documented now. Still missing is an update of the README, but I will leave that for later, because I want to cut another RC release to "test" Steramee against Stop Resolver (permanent) and Dispatching 2 (perRequest).

**Pull request merged**